[receiver/sqlquery] add support for logs

andrzej-stencel opened the pull request:
Fixes #20284 

This introduces initial support for retrieving rows from SQL databases into logs.

This PR aims to provide an initial, not feature rich, but production ready implementation. The following features are available:

- Use `body_column` to select the column to use to fill the Body field of the created log
- Use `tracking_start_value` and `tracking_column` properties to track rows that were already ingested
- Use `storage` property to persist the tracking value across collector restarts

In this state and marked as "development" stability, the component can be used for experimentation and to guide future development.

There are definitely more things that need to be implemented for this component to be considered "alpha" quality - like filling in other [log fields](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.21.0/specification/logs/data-model.md#log-and-event-record-definition) like Timestamp, ObservedTimestamp and others. I would like to add them in subsequent pull requests, as this pull request is already way too big.

andrzej-stencel committed:
feat: add config option for logs

Adds basic configuration option for logs.

andrzej-stencel committed:
ci: add changelog entry

andrzej-stencel committed:
style: apply Prettier's fixes in README.md

atoulme commented on receiver/sqlqueryreceiver/README.md:
```diff
@@ -1,7 +1,7 @@
-|--------------------------|-----------|
+| ------------------------ | --------- |
 | Stability                | [alpha]   |
 | Supported pipeline types | metrics   |
```
```suggestion
| Supported pipeline types | logs, metrics   |
```
Would it be accurate to say you'd want to have this receiver support log pipelines?

astencel-sumo replied:
Yes, of course, thanks Antoine :slightly_smiling_face:

andrzej-stencel committed:
docs: add logs support to the status table

andrzej-stencel committed:
test: add integration tests for reading logs

The integration tests are failing now,
because log collection is not implemented yet.
Implementing it is the next step.
The initial implementation in the next step should be the most naive possible:
just running the provided query every collection interval,
without any tracking of the data that has already been read.

Perhaps the log collection integration tests
should be separated from the metrics collection
integration tests for greater visibility
into what is working and what is failing.

dmolenda-sumo committed:
scrape logs

dmolenda-sumo committed:
read simple logs from database

andrzej-stencel committed:
feat: panic on query errors from db

The panic is useful for testing.
When this feature is complete,
the panics must be replaced with error logging.

andrzej-stencel committed:
test: fix Oracle integration test

andrzej-stencel committed:
test(integration): expect 5 log records

At this stage, the logs query is run only once
per collector run, e.g. collection_interval is ignored.
Reflect this in the tests before implementing
collecting logs on interval.

andrzej-stencel committed:
feat: make body column configurable

Oracle expects uppercase column names.

andrzej-stencel committed:
feat: collect logs periodically

Collecting logs every collection interval.
To prevent collecting duplicates,
we need to track the already collected logs,
which is going to be the next step.

andrzej-stencel committed:
Merge branch 'main' into sqlquery-receiver-add-logs

andrzej-stencel committed:
refactor: go mod tidy

andrzej-stencel committed:
feat: support tracking processed rows

This initial implementation supports only integer tracking columns.
The next type to support should be datetime/timestamp.
We should also support using persistent storage for the tracking value.

andrzej-stencel committed:
fix: do not try to store tracking value when not configured

andrzej-stencel committed:
fix: initialize shutdown channel

andrzej-stencel committed:
fix: do not send empty logs data

andrzej-stencel committed:
Merge branch 'main' into sqlquery-receiver-add-logs

dmolenda-sumo committed:
SQL Query receiver - support timestamp in tracking column

dmolenda-sumo committed:
Merge pull request #2 from dmolenda-sumo/sqlqueryreceiver-support-timestamp-for-tracking-column

SQL Query receiver - support timestamp in tracking column

dmolenda-sumo committed:
Merge remote-tracking branch 'upstream/main' into sqlquery-receiver-add-logs

dmolenda-sumo committed:
fix integration tests

dmolenda-sumo committed:
Merge pull request #3 from dmolenda-sumo/sqlqueryreceiver-fix-integration-tests

fix integration tests

kasia-kujawa committed:
feat: add observability (metrics) for sqlquery receiver (#1)

andrzej-stencel committed:
feat: persist tracking value in storage

dmolenda-sumo committed:
SQL Query receiver: add logs to mysql integration test

dmolenda-sumo committed:
Merge pull request #5 from dmolenda-sumo/sqlqueryreceiver-add-logs-to-mysql-integration-test

SQL Query receiver: add logs to mysql integration test

kasia-kujawa committed:
feat(sqlqueryreceiver) add error count metrics for receiver (#4)

andrzej-stencel committed:
docs: add link to a storage extension

andrzej-stencel committed:
style: remove unused code

andrzej-stencel committed:
test: terminate DB containers after test

This prevents a "busy port" error when running
one of the tests repeatedly.

andrzej-stencel committed:
Merge branch 'sqlquery-receiver-add-logs' into persistence

andrzej-stencel committed:
Merge pull request #6 from dmolenda-sumo/persistence

feat: persist tracking value in storage

andrzej-stencel committed:
Merge branch 'main' into sqlquery-receiver-add-logs

andrzej-stencel committed:
test: fix after metadata update

andrzej-stencel committed:
refactor: use metadata for logs stability level

andrzej-stencel committed:
Merge branch 'main' into sqlquery-receiver-add-logs

andrzej-stencel committed:
docs: add note about storage

andrzej-stencel committed:
docs: fix parameter notation in configuration

The `$` character needs to be escaped with `$$` in Otelcol config files.

andrzej-stencel committed:
chore: go mod tidy

andrzej-stencel committed:
fix lint errors

andrzej-stencel committed:
Merge branch 'main' into sqlquery-receiver-add-logs

andrzej-stencel committed:
chore: update license to short form

andrzej-stencel committed:
chore: make crosslink

andrzej-stencel committed:
chore: make gotidy

kasia-kujawa committed:
feat(sqlqueryreceiver) add traces to monitor logs received

andrzej-stencel committed:
docs: disable Prettier formatting to pass CI check

andrzej-stencel committed:
chore: make generate

andrzej-stencel committed:
Merge pull request #7 from kkujawa-sumo/kk-add-traces

feat(sqlqueryreceiver) add traces to monitor logs received

andrzej-stencel committed:
Merge branch 'main' into sqlquery-receiver-add-logs

dmitryax commented on receiver/sqlqueryreceiver/internal/observability/observability.go:
```diff
@@ -0,0 +1,88 @@
+)
+
+var (
+	mAcceptedLogs = stats.Int64("receiver/accepted/log/records", "Number of log record pushed into the pipeline.", "")
+	mErrorCount   = stats.Int64("receiver/errors", "Number of errors", "")
```
I believe we should use standard metrics for the consumers instead. These metrics seem to serve the same purpose. Please see https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/1d7655100b1bd42ab7863e70cf8abc31aac03da8/pkg/stanza/adapter/receiver.go#L123-L128 as example

dmitryax replied:
Why will refused logs never be reported?

astencel-sumo replied:
When doing `nextConsumer.ConsumeLogs()` ([source](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/20730/files#diff-0c9a67fda52566bc14742cf772b15a23ac8c74c66e5fc4f9b86c734b809d4a09R181)), the db call error is not available, as the error is specific to a particular query, and there can be more than one query configured for the receiver.

If you look in the logs receiver's `collect()` method ([source](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/20730/files#diff-0c9a67fda52566bc14742cf772b15a23ac8c74c66e5fc4f9b86c734b809d4a09R147)), it concurrently runs all the db queries (using a "query receiver" for each query) and passes all the collected logs to the next consumer. If an error occurs in any of the "query receivers", it is logged and dropped, and the returned log count is zero.

Not sure if this should be structured differently? I don't think each "query receiver" should run `nextConsumer.ConsumeLogs()` on their own? :thinking:

astencel-sumo replied:
@dmitryax I have removed the duplicate metrics, only using the `EndLogsOp` function now. Let's get the PR merged and iron out the metrics details in a next step?

andrzej-stencel committed:
Merge branch 'main' into sqlquery-receiver-add-logs

andrzej-stencel committed:
fix integration test

andrzej-stencel committed:
test: clean up tests

After introducing integration test framework

andrzej-stencel committed:
fix: prevent panic on shutdown

When the component needs to be shut down before it was properly started
(for example due to a configuration erro
like using a storage extension that does not exist),
we need to check if variables have been initialized
to prevent panics due to nil pointer dereference.

andrzej-stencel committed:
Merge branch 'main' into sqlquery-receiver-add-logs

andrzej-stencel committed:
remove duplicate metrics

andrzej-stencel committed:
Merge branch 'main' into sqlquery-receiver-add-logs

dmitryax commented on receiver/sqlqueryreceiver/logs_receiver.go:
```diff
@@ -0,0 +1,318 @@
+	scopeLogs := logs.ResourceLogs().AppendEmpty().ScopeLogs().AppendEmpty().LogRecords()
+	for logsConfigIndex, logsConfig := range queryReceiver.query.Logs {
+		for i, row := range rows {
+			if err = rowToLog(row, logsConfig, scopeLogs.AppendEmpty()); err != nil {
```
`rowToLog` doesn't return any error. Let's clean up the redundant error handling?

astencel-sumo replied:
OK, removed. I had left the error checking in there because I expected the `rowToLog` function to return errors in the future, but yeah, we can add error checking when it's needed.

andrzej-stencel committed:
refactor: remove unused error path

dmitryax commented:
Please rebase

andrzej-stencel committed:
Merge branch 'main' into sqlquery-receiver-add-logs

astencel-sumo commented:
Done @dmitryax
